File: package/node/src/modules/totp.ts
import type { Transport } from "../core/transport";
import type { SessionManager } from "../core/session";
import type {
  TOTPConfig,
  TOTPSetupResponse,
  TOTPStatusResponse,
  TOTPVerifyInput,
  TOTPLoginInput,
  TokenResponse,
} from "../types";
import { SessionExpiredError } from "../errors";

interface TOTPModuleDeps {
  transport: Transport;
  session: SessionManager;
  config?: TOTPConfig;
}

class TOTPModule {
  private transport: Transport;
  private session: SessionManager;
  private config?: TOTPConfig;

  constructor(deps: TOTPModuleDeps) {
    this.transport = deps.transport;
    this.session = deps.session;
    this.config = deps.config;
  }

  async setup(): Promise<TOTPSetupResponse> {
    const accessToken = this.session.getAccessToken();

    if (!accessToken) {
      throw new SessionExpiredError();
    }

    const headers: Record<string, string> = {};
    if (this.config?.issuer) {
      headers["X-TOTP-Issuer"] = this.config.issuer;
    }

    return this.transport.post<TOTPSetupResponse>(
      "/api/v1/auth/totp/setup",
      undefined,
      accessToken,
    );
  }

  async verify(input: TOTPVerifyInput): Promise<void> {
    const accessToken = this.session.getAccessToken();

    if (!accessToken) {
      throw new SessionExpiredError();
    }

    await this.transport.post(
      "/api/v1/auth/totp/verify",
      {
        code: input.code,
        secret: input.secret,
      },
      accessToken,
    );
  }

  async disable(): Promise<void> {
    const accessToken = this.session.getAccessToken();

    if (!accessToken) {
      throw new SessionExpiredError();
    }

    await this.transport.post(
      "/api/v1/auth/totp/disable",
      undefined,
      accessToken,
    );
  }

  async getStatus(): Promise<TOTPStatusResponse> {
    const accessToken = this.session.getAccessToken();

    if (!accessToken) {
      throw new SessionExpiredError();
    }

    return this.transport.get<TOTPStatusResponse>(
      "/api/v1/auth/totp/status",
      accessToken,
    );
  }

  async login(
    input: TOTPLoginInput,
    oauthParams?: OAuthParams,
  ): Promise<TokenResponse> {
    const endpoint = this.buildLoginEndpoint(oauthParams);

    const response = await this.transport.post<TokenResponse>(endpoint, {
      email: input.email,
      password: input.password,
      totpCode: input.totpCode,
    });

    this.session.setTokens(response);
    return response;
  }

  private buildLoginEndpoint(oauthParams?: OAuthParams): string {
    if (!oauthParams || Object.keys(oauthParams).length === 0) {
      return "/api/v1/auth/totp/login";
    }

    const params = new URLSearchParams();
    if (oauthParams.client_id) params.set("client_id", oauthParams.client_id);
    if (oauthParams.redirect_uri)
      params.set("redirect_uri", oauthParams.redirect_uri);
    if (oauthParams.response_type)
      params.set("response_type", oauthParams.response_type);
    if (oauthParams.scope) params.set("scope", oauthParams.scope);
    if (oauthParams.state) params.set("state", oauthParams.state);

    return `/api/v1/auth/totp/login?${params.toString()}`;
  }
}

interface OAuthParams {
  client_id?: string;
  redirect_uri?: string;
  response_type?: string;
  scope?: string;
  state?: string;
}

export { TOTPModule };
export type { OAuthParams };

File: package/python/aether_identity/modules/totp.py
"""Aether Identity Python SDK - TOTP module."""

from typing import Optional
from ..core.transport import Transport
from ..core.session import SessionManager
from ..types import (
    TOTPSetupResponse,
    TOTPVerifyInput,
    TOTPStatusResponse,
    TOTPLoginInput,
    TokenResponse,
    OAuthParams,
)


class TOTPModule:
    """TOTP/2FA operations module."""

    def __init__(self, transport: Transport, session_manager: SessionManager):
        """Initialize TOTP module.

        Args:
            transport: HTTP transport instance
            session_manager: Session manager instance
        """
        self._transport = transport
        self._session = session_manager

    def setup(self) -> TOTPSetupResponse:
        """Setup TOTP for user.

        Returns:
            TOTP setup data with secret and QR code
        """
        response = self._transport.post(
            "/api/v1/auth/totp/setup", access_token=self._session.get_access_token()
        )

        return TOTPSetupResponse.model_validate(response)

    def verify(self, input_data: TOTPVerifyInput) -> dict:
        """Verify TOTP code.

        Args:
            input_data: TOTP verification input

        Returns:
            API response
        """
        data = input_data.model_dump(exclude_none=True, by_alias=True)

        return self._transport.post(
            "/api/v1/auth/totp/verify", data, access_token=self._session.get_access_token()
        )

    def disable(self) -> dict:
        """Disable TOTP.

        Returns:
            API response
        """
        return self._transport.post(
            "/api/v1/auth/totp/disable", access_token=self._session.get_access_token()
        )

    def get_status(self) -> TOTPStatusResponse:
        """Check TOTP status.

        Returns:
            TOTP status response
        """
        response = self._transport.get(
            "/api/v1/auth/totp/status", access_token=self._session.get_access_token()
        )

        return TOTPStatusResponse.model_validate(response)

    def login(
        self, input_data: TOTPLoginInput, oauth_params: Optional[OAuthParams] = None
    ) -> TokenResponse:
        """Login with TOTP.

        Args:
            input_data: TOTP login input
            oauth_params: Optional OAuth parameters

        Returns:
            Token response
        """
        data = input_data.model_dump(by_alias=True)

        if oauth_params:
            data.update(oauth_params.model_dump(exclude_none=True, by_alias=True))

        response = self._transport.post("/api/v1/auth/totp/login", data)

        tokens = TokenResponse.model_validate(response)
        self._session.set_tokens(tokens)

        return tokens
